components/event-sources/adapter/http/adapter.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/kyma-project/kyma/components/event-sources/apis/sources"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/adapter"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/source"
+	"log"
+	"net/http"
+)
+
+type envConfig struct {
+	adapter.EnvConfig
+	EventSource string `envconfig:"EVENT_SOURCE" required:"true"`
+
+	// PORT as required by knative serving runtime contract
+	Port int `envconfig:"PORT" required:"true" default:"8080"`
+}
+
+func (e *envConfig) GetSource() string {
+	return e.EventSource
+}
+
+func (e *envConfig) GetPort() int {
+	return e.Port
+}
+
+type httpAdapter struct {
+	ceClient       cloudevents.Client
+	statsReporter  source.StatsReporter
+	accessor       AdapterEnvConfigAccessor
+	adapterContext context.Context
+	logger         *zap.Logger
+}
+
+type AdapterEnvConfigAccessor interface {
+	adapter.EnvConfigAccessor
+	GetSource() string
+	GetPort() int
+}
+
+const resourceGroup = "http." + sources.GroupName
+
+const (
+	// endpoint for cloudevents
+	endpointCE = "/"
+	// endpoint for readiness check
+	readinessReadiness = "/healthz"
+)
+
+const (
+	ErrorResponseCEVersionUnsupported = "unsupported cloudevents version"
+	ErrorResponseSendToSinkFailed     = "unable to forward event to sink"
+)
+
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &envConfig{}
+}
+
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client, reporter source.StatsReporter) adapter.Adapter {
+
+	accessor, ok := processed.(AdapterEnvConfigAccessor)
+
+	if !ok {
+		panic(fmt.Sprintf("cannot create adapter, expecting a *envconfig, but got a %T", processed))
+	}
+
+	return &httpAdapter{
+		adapterContext: ctx,
+		ceClient:       ceClient,
+		statsReporter:  reporter,
+		accessor:       accessor,
+		logger:         logging.FromContext(ctx).Desugar(),
+	}
+}
+
+// Start is the entrypoint for the adapter and is called by sharedmain coming from pkg/adapter
+func (h *httpAdapter) Start(_ <-chan struct{}) error {
+	logger := h.logger
+
+	t, err := cloudevents.NewHTTPTransport(
+		cloudevents.WithPort(h.accessor.GetPort()),
+		cloudevents.WithPath(endpointCE),
+		cloudevents.WithMiddleware(WithReadinessMiddleware),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create transport, %v", err)
+	}
+
+	c, err := cloudevents.NewClient(t)
+	if err != nil {
+		return fmt.Errorf("failed to create client, %v", err)
+	}
+
+	log.Printf("listening on :%d%s\n", h.accessor.GetPort(), endpointCE)
+
+	// note about graceful shutdown:
+	// TLDR; StartReceiver unblocks as soon as a stop signal is received
+	// `StartReceiver` waits internally until `ctx.Done()` does not block anymore
+	// the context `h.adapterContext` returns a channel (when calling `ctx.Done()`)
+	// which is closed as soon as a stop signal is received, see https://github.com/knative/pkg/blob/master/signals/signal.go#L37
+	if err := c.StartReceiver(h.adapterContext, h.serveHTTP); err != nil {
+		return fmt.Errorf("failed to start receiver: %v", err)
```

Comment:
The error  message is misleading, an error could also occur after the receiver has been successfully started if I'm not mistaken.

Reply:
True.